internal/services/ingest.go

```diff
@@ -155,6 +169,212 @@ func (m *ingestService) Run(ctx context.Context, startLedger uint32, endLedger u
+		totalIngestionStart := time.Now()
+		// process ledgers
+		err = m.processLedgerResponse(ctx, getLedgersResponse)
+		if err != nil {
+			return fmt.Errorf("processing ledger response: %w", err)
+		}
+
+		// update cursor
+		startLedger = getLedgersResponse.Ledgers[len(getLedgersResponse.Ledgers)-1].Sequence
+		if err := m.models.IngestStore.UpdateLatestLedgerSynced(ctx, m.ledgerCursorName, startLedger); err != nil {
+			return fmt.Errorf("updating latest synced ledger: %w", err)
+		}
+		m.metricsService.SetLatestLedgerIngested(float64(getLedgersResponse.LatestLedger))
+		m.metricsService.ObserveIngestionDuration(totalIngestionPrometheusLabel, time.Since(totalIngestionStart).Seconds())
+
+		if len(getLedgersResponse.Ledgers) == getLedgersLimit {
+			ticker.Reset(tickerDuration)
+			manualTriggerChan <- nil
+		}
+	}
+}
+
+// fetchNextLedgersBatch fetches the next batch of ledgers from the RPC service.
+func (m *ingestService) fetchNextLedgersBatch(ctx context.Context, startLedger uint32) (GetLedgersResponse, error) {
+	rpcHealth, err := m.rpcService.GetHealth()
+	if err != nil {
+		return GetLedgersResponse{}, fmt.Errorf("getting rpc health: %w", err)
+	}
+	ledgerSeqRange, inSync := getLedgerSeqRange(rpcHealth.OldestLedger, rpcHealth.LatestLedger, startLedger)
+	log.Ctx(ctx).Debugf("ledgerSeqRange: %+v", ledgerSeqRange)
+	if inSync {
+		return GetLedgersResponse{}, ErrAlreadyInSync
+	}
+
+	getLedgersResponse, err := m.rpcService.GetLedgers(ledgerSeqRange.StartLedger, ledgerSeqRange.Limit)
+	if err != nil {
+		return GetLedgersResponse{}, fmt.Errorf("getting ledgers: %w", err)
+	}
+
+	return getLedgersResponse, nil
+}
+
+type LedgerSeqRange struct {
+	StartLedger uint32
+	Limit       uint32
+}
+
+// getLedgerSeqRange returns a ledger sequence range to ingest. It takes into account:
+// - the ledgers available in the RPC,
+// - the latest ledger synced by the ingestion service,
+// - the max ledger window to ingest.
+//
+// The returned ledger sequence range is inclusive of the start and end ledgers.
+func getLedgerSeqRange(rpcOldestLedger, rpcNewestLedger, latestLedgerSynced uint32) (ledgerRange LedgerSeqRange, inSync bool) {
+	if latestLedgerSynced >= rpcNewestLedger {
+		return LedgerSeqRange{}, true
+	}
+	ledgerRange.StartLedger = max(latestLedgerSynced+1, rpcOldestLedger)
+	ledgerRange.Limit = getLedgersLimit
+
+	return ledgerRange, false
+}
+
+func (m *ingestService) processLedgerResponse(ctx context.Context, getLedgersResponse GetLedgersResponse) error {
+	// Create a worker pool with
+	const poolSize = 16
+	pool := pond.New(poolSize, len(getLedgersResponse.Ledgers), pond.Context(ctx))
+
+	// Create a slice of errors
+	mu := sync.Mutex{}
+	var allTxHashes []string
+	var errs []error
+
+	// Submit tasks to the pool
+	for _, ledger := range getLedgersResponse.Ledgers {
+		ledger := ledger // Create a new variable to avoid closure issues
+		pool.Submit(func() {
+			txHashes, err := m.processLedger(ctx, ledger)
+			mu.Lock()
+			if err != nil {
```

Comment:
I plan to use toid to order the transactions in the database. It's calculated like [this](https://github.com/stellar/go/blob/ef10cd0a592065357e7dff199401c38494b461bf/processors/operation/operation.go#L49).